server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java

```diff
@@ -953,11 +928,116 @@ protected void stopInternal() {}
             verify(httpChannel).sendResponse(response.capture(), listener.capture());
 
             listener.getValue().onResponse(null);
-            assertThat(response.getValue().headers().get(DefaultRestChannel.CONNECTION), containsInAnyOrder(DefaultRestChannel.CLOSE));
+            assertThat(response.getValue().headers().get(CONNECTION), containsInAnyOrder(DefaultRestChannel.CLOSE));
             verify(httpChannel).close();
         }
     }
 
+    public void testStopDoesntWaitIfGraceIsZero() {
+        try (TestHttpServerTransport transport = new TestHttpServerTransport(Settings.EMPTY)) {
+            transport.bindServer();
+            final TestHttpRequest httpRequest = new TestHttpRequest(HttpRequest.HttpVersion.HTTP_1_1, RestRequest.Method.GET, "/");
+
+            TestHttpChannel httpChannel = new TestHttpChannel(false);
+            transport.serverAcceptedChannel(httpChannel);
+            transport.incomingRequest(httpRequest, httpChannel);
+
+            transport.doStop();
+            assertThat("stop listening first", transport.stopListeningForNewConnectionsOrder(), is(1));
```

Comment:
Relying on the exact order internal methods are called seems fragile. Could we instead take a more black box approach to the transport by testing the behavior we expect? That is eg open a new connection with the transport, but mock out the dispatcher so that you can hang the response when desired.  I realize it's more complicated than that in practice, but I think testing method calls is relying too heavily on low level implementation details which will make any future modifications to this code difficult.

Reply:
Removed the helper functions.  The only one I kept was `gracefullyCloseConnections()` which was already there (but I moved it closer to it's use in `doStop()`).  The tests need that hook, still.

Reply:
Everything else is doing black box testing now.